### PR TITLE
Avoid Slice allocation in SliceWriters

### PR DESCRIPTION
**Pull request by arunthirupathi**
There are two paths where Slices are allocated in Slice Writers.
1. Statistics Calculation requires Slices to compute min/max/sum. 
The first PR calculates the Statistics directly from block and
position. This avoids the Slice allocation.
2. Writing to OrcOutputBuffer requires Slice.  The second PR
appends the block position directly to the OrcOutputBuffer.

SliceWriters after this change, will not allocate short lived Slices.

Test plan 
Existing tests.
Ran Validation Service and cross reader/writer validation.

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

**Comment by sdruzkin** [comment]
Lol, I just did exactly the same change to minimize slice allocation in statistics. I did it and tested directly with block w/o the type.

I even did benchmarking and was going to post a PR:
```
B - Before, A - After

   Benchmark                                           (typeSignature)  (uniqueValuesPercentage)  Mode  Cnt     Score    Error  Units
B: BenchmarkDictionaryWriter.writeDictionaryNoSorting          varchar                         1  avgt   50   983.346 ± 19.763  ms/op
A: BenchmarkDictionaryWriter.writeDictionaryNoSorting          varchar                         1  avgt   50   820.430 ± 12.343  ms/op

B: BenchmarkDictionaryWriter.writeDictionaryNoSorting          varchar                         5  avgt   50  1240.838 ± 25.541  ms/op
A: BenchmarkDictionaryWriter.writeDictionaryNoSorting          varchar                         5  avgt   50  1080.936 ± 21.458  ms/op

B: BenchmarkDictionaryWriter.writeDictionaryNoSorting          varchar                        10  avgt   50  1536.781 ± 23.302  ms/op
A: BenchmarkDictionaryWriter.writeDictionaryNoSorting          varchar                        10  avgt   50  1316.958 ± 24.234  ms/op


B: BenchmarkDictionaryWriter.writeDictionaryNoSorting          varchar                        25  avgt   50  2592.528 ± 45.479  ms/op
A: BenchmarkDictionaryWriter.writeDictionaryNoSorting          varchar                        25  avgt   50  2361.023 ± 34.907  ms/op

B: BenchmarkDictionaryWriter.writeDictionaryNoSorting          varchar                        50  avgt   50  3283.906 ± 35.623  ms/op
A: BenchmarkDictionaryWriter.writeDictionaryNoSorting          varchar                        50  avgt   50  3095.365 ± 28.988  ms/op

B: BenchmarkDictionaryWriter.writeDictionaryNoSorting          varchar                       100  avgt   50  5097.326 ± 45.289  ms/op
A: BenchmarkDictionaryWriter.writeDictionaryNoSorting          varchar                       100  avgt   50  4804.164 ± 80.545  ms/op
```

**Comment by sdruzkin** [comment]
Here is my version https://github.com/prestodb/presto/pull/17362

**Comment by arunthirupathi** [comment]
> Here is my version #17362

Both code looks like they are plagiarized from each other :) . Apologies, did not know you were working on this, I was planning to wrap up the writer work after the CPU scare from Spark rollout.

File: presto-orc/src/test/java/com/facebook/presto/orc/metadata/statistics/TestBinaryStatisticsBuilder.java [hunk 1]
```diff
@@ -122,4 +112,11 @@ private void assertBinaryStatistics(ColumnStatistics columnStatistics, int expec
             assertEquals(columnStatistics.getNumberOfValues(), 0);
         }
     }
+
+    public static void addValue(BinaryStatisticsBuilder binaryStatisticsBuilder, Slice slice)
+    {
+        int[] offsets = new int[] {0, slice.length()};
```
Comment (sdruzkin) [inline presto-orc/src/test/java/com/facebook/presto/orc/metadata/statistics/TestBinaryStatisticsBuilder.java:118]: This won't work with nulls, right?

**Comment by sdruzkin** [comment]
> > Here is my version #17362
> 
> Both code looks like they are plagiarized from each other :) . Apologies, did not know you were working on this, I was planning to wrap up the writer work after the CPU scare from Spark rollout.

Me is the one who should apologize while working being on vacation, lol.

What do you think about dumping the type for the statistics and making it work directly with blocks?

File: presto-orc/src/test/java/com/facebook/presto/orc/metadata/statistics/TestBinaryStatisticsBuilder.java (hunk 1 above)
Comment (arunthirupathi) [reply presto-orc/src/test/java/com/facebook/presto/orc/metadata/statistics/TestBinaryStatisticsBuilder.java:118]: OrcWriter for statistics does not track nulls. It only tracks non nulls explicitly so this is fine.

**Comment by arunthirupathi** [comment]
> > > Here is my version #17362
> > 
> > 
> > Both code looks like they are plagiarized from each other :) . Apologies, did not know you were working on this, I was planning to wrap up the writer work after the CPU scare from Spark rollout.
> 
> Me is the one who should apologize while working being on vacation, lol.
> 
> What do you think about dumping the type for the statistics and making it work directly with blocks?

We can dump types and that would make it consistent. Types handling are very inconsistent now.

Contents of the data, does not use type to get slices.
Statistics uses types for comparison. 

Please let me know if we should get rid of types from Statistics as well.

**Comment by sdruzkin** [comment]
Btw, there is an opportunity to check whether the value was really added to the dictionary to avoid that comparison for min/max in stats. But the problem is that we would reset stats for every row group, so it need additional buffer per row group to remember whether the value is seen in the current row group for the first time or not.

**Comment by arunthirupathi** [comment]
> Btw, there is an opportunity to check whether the value was really added to the dictionary to avoid that comparison for min/max in stats. But the problem is that we would reset stats for every row group, so it needs an additional buffer to remember whether the value is seen in the current row group for the first time or not.

I removed this comment for this exact reason, we need some per row group tracker. So the comment was not right.

One thing we can do is, when we abandon dictionary encoding and switch to direct we can port the statistics instead of recomputing it again. But the problem is last open row group needs re computation. So the API is kind of complicated and was not showing that big of a win.

File: presto-orc/src/main/java/com/facebook/presto/orc/writer/SliceDirectColumnWriter.java [hunk 2]
```diff
@@ -85,11 +86,12 @@ public SliceDirectColumnWriter(
             MetadataWriter metadataWriter)
     {
         checkArgument(column >= 0, "column is negative");
+        checkArgument(type instanceof AbstractVariableWidthType, "type is not a AbstractVariableWidthType");
```
Comment (sdruzkin) [inline presto-orc/src/main/java/com/facebook/presto/orc/writer/SliceDirectColumnWriter.java:89]: type is not `a` AbstractVariableWidthType -> type is not `an` AbstractVariableWidthType

File: presto-orc/src/test/java/com/facebook/presto/orc/metadata/statistics/TestBinaryStatisticsBuilder.java [hunk 3]
```diff
@@ -122,4 +112,11 @@ private void assertBinaryStatistics(ColumnStatistics columnStatistics, int expec
...
+        VariableWidthBlock block = new VariableWidthBlock(1, slice, offsets, Optional.empty());
```
Comment (sdruzkin) [inline presto-orc/src/test/java/com/facebook/presto/orc/metadata/statistics/TestBinaryStatisticsBuilder.java:119]: I'd prefer to use the builder to avoid any unnecessary changes in case if VariableWidthBlock constructor is changed.

File: presto-orc/src/test/java/com/facebook/presto/orc/writer/TestSegmentedSliceBlockBuilder.java [hunk 4]
```diff
@@ -89,6 +95,33 @@ public void testCompareTo()
...
     }
 
+    @Test
+    public void testWriteBytes()
+    {
+        int entries = 100;
+        String inputChars = "abcdefghijklmnopqrstuvwwxyz01234566789!@#$%^";
+        SegmentedSliceBlockBuilder blockBuilder = new SegmentedSliceBlockBuilder(entries, inputChars.length());
+        List<String> values = new ArrayList<>();
+        Random rand = new Random();
```
Comment (sdruzkin) [inline presto-orc/src/test/java/com/facebook/presto/orc/writer/TestSegmentedSliceBlockBuilder.java:105]: use seeded random

File: presto-common/src/main/java/com/facebook/presto/common/type/AbstractVarcharType.java [hunk 5]
```diff
@@ -105,6 +106,19 @@ public int compareTo(Block leftBlock, int leftPosition, Block rightBlock, int ri
         return leftBlock.compareTo(leftPosition, 0, leftLength, rightBlock, rightPosition, 0, rightLength);
     }
 
+    @Override
+    public int compareTo(Block leftBlock, int leftPosition, Slice rightSlice, int sliceOffset, int sliceLength)
```
[the remark this answers is not in the document]
Comment (arunthirupathi) [reply presto-common/src/main/java/com/facebook/presto/common/type/AbstractVarcharType.java:110]: I did verifier run for this change but I will remove the Type changes.

File: presto-orc/src/main/java/com/facebook/presto/orc/metadata/statistics/StringStatisticsBuilder.java [hunk 6]
```diff
@@ -61,25 +62,23 @@ public long getNonNullValueCount()
     }
 
     @Override
-    public void addValue(Slice value, int sourceIndex, int length)
+    public void addValue(Block block, int position)
     {
-        requireNonNull(value, "value is null");
-
+        int sliceLength = block.getSliceLength(position);
```
Comment (sdruzkin) [inline presto-orc/src/main/java/com/facebook/presto/orc/metadata/statistics/StringStatisticsBuilder.java:67]: Sprinkle a bit of requireNonNull?

File: presto-orc/src/main/java/com/facebook/presto/orc/metadata/statistics/BinaryStatisticsBuilder.java [hunk 7]
```diff
@@ -28,10 +28,9 @@
     private long sum;
 
     @Override
-    public void addValue(Slice value, int sourceIndex, int length)
+    public void addValue(Block block, int position)
     {
-        requireNonNull(value, "value is null");
-        sum += length;
+        sum += block.getSliceLength(position);
```
Comment (sdruzkin) [inline presto-orc/src/main/java/com/facebook/presto/orc/metadata/statistics/BinaryStatisticsBuilder.java:33]: Sprinkle a bit of requireNonNull?

File: presto-orc/src/test/java/com/facebook/presto/orc/writer/TestSegmentedSliceBlockBuilder.java [hunk 8]
```diff
@@ -89,6 +95,33 @@ public void testCompareTo()
...
 
+    @Test
+    public void testWriteBytes()
+    {
+        int entries = 100;
+        String inputChars = "abcdefghijklmnopqrstuvwwxyz01234566789!@#$%^";
+        SegmentedSliceBlockBuilder blockBuilder = new SegmentedSliceBlockBuilder(entries, inputChars.length());
+        List<String> values = new ArrayList<>();
+        Random rand = new Random(0);
+        byte[] bytes = inputChars.getBytes(StandardCharsets.UTF_8);
```
Comment (sdruzkin) [inline presto-orc/src/test/java/com/facebook/presto/orc/writer/TestSegmentedSliceBlockBuilder.java:106]: Static import for UTF_8

**Review by sdruzkin — APPROVED** [review]
LGTM

File: presto-orc/src/test/java/com/facebook/presto/orc/metadata/statistics/TestBinaryStatisticsBuilder.java [hunk 9]
```diff
@@ -53,24 +55,11 @@ public void testSum()
...
-        BinaryStatisticsBuilder binaryStatisticsBuilder = new BinaryStatisticsBuilder();
-        Slice slice = utf8Slice("abcdefghijklmnopqrstuvwxyz");
-        for (int i = 0; i < slice.length(); i++) {
-            binaryStatisticsBuilder.addValue(slice, i, 1);
-        }
-
-        BinaryStatistics binaryStatistics = binaryStatisticsBuilder.buildColumnStatistics().getBinaryStatistics();
-        assertEquals(binaryStatistics.getSum(), slice.length());
-    }
-
```
Comment (NikhilCollooru) [inline presto-orc/src/test/java/com/facebook/presto/orc/metadata/statistics/TestBinaryStatisticsBuilder.java:73]: Shall we add something similar for block ? also in TestStringStatisticsBuilder ?

File: presto-common/src/main/java/com/facebook/presto/common/block/Block.java [hunk 10]
```diff
@@ -120,6 +120,11 @@ default void writeBytesTo(int position, int offset, int length, BlockBuilder blo
         throw new UnsupportedOperationException(getClass().getName());
     }
```
Comment (NikhilCollooru) [inline presto-common/src/main/java/com/facebook/presto/common/block/Block.java:122]: Can we add a comment similar to above

File: presto-common/src/main/java/com/facebook/presto/common/block/AbstractSingleArrayBlock.java [hunk 11]
```diff
@@ -114,6 +114,12 @@ public void writeBytesTo(int position, int offset, int length, BlockBuilder bloc
         getBlock().writeBytesTo(position + start, offset, length, blockBuilder);
     }
 
+    public void writeBytesTo(int position, int offset, int length, SliceOutput sliceOutput)
```
Comment (highker) [inline presto-common/src/main/java/com/facebook/presto/common/block/AbstractSingleArrayBlock.java:117]: `@Override`

File: presto-common/src/main/java/com/facebook/presto/common/block/AbstractSingleMapBlock.java [hunk 12]
```diff
@@ -141,6 +141,17 @@ public void writeBytesTo(int position, int offset, int length, BlockBuilder bloc
         }
     }
 
+    public void writeBytesTo(int position, int offset, int length, SliceOutput sliceOutput)
```
Comment (highker) [inline presto-common/src/main/java/com/facebook/presto/common/block/AbstractSingleMapBlock.java:144]: `@Override`

File: presto-common/src/main/java/com/facebook/presto/common/type/AbstractVarcharType.java [hunk 13]
```diff
@@ -146,6 +147,11 @@ public void writeSlice(BlockBuilder blockBuilder, Slice value, int offset, int l
         blockBuilder.writeBytes(value, offset, length).closeEntry();
     }
 
+    public void writeBytes(AbstractVariableWidthBlockBuilder blockBuilder, byte[] bytes, int offset, int length)
+    {
+        blockBuilder.writeBytes(bytes, offset, length).closeEntry();
+    }
```
Comment (highker) [inline presto-common/src/main/java/com/facebook/presto/common/type/AbstractVarcharType.java:153]: Is this only used in tests? If yes, maybe move to test-only classes?

**Review by highker — APPROVED** [review]

File: presto-common/src/main/java/com/facebook/presto/common/type/AbstractVarcharType.java (hunk 13 above)
Comment (arunthirupathi) [reply presto-common/src/main/java/com/facebook/presto/common/type/AbstractVarcharType.java:153]: This is going to be used in Jolly , Presto will not use it for now.
